From 8e4223e9084a32a3dee92cb5e3aebf140e180423 Mon Sep 17 00:00:00 2001
From: Robert Hodaszi <robert.hodaszi@digi.com>
Date: Mon, 25 Nov 2019 17:44:52 +0100
Subject: [PATCH] Add survey data fallback, instead of failing immediately

Some drivers (like mt76) are missing certain survey data (mt76 -> noise
floor).

Pulled from OpenWRT

Signed-off-by: Robert Hodaszi <robert.hodaszi@digi.com>
---
 src/ap/acs.c | 13 +++----------
 1 file changed, 3 insertions(+), 10 deletions(-)

diff --git a/src/ap/acs.c b/src/ap/acs.c
index 6d4c041..54925fd 100644
--- a/src/ap/acs.c
+++ b/src/ap/acs.c
@@ -302,16 +302,12 @@ static void acs_fail(struct hostapd_iface *iface)
 static long double
 acs_survey_interference_factor(struct freq_survey *survey, s8 min_nf)
 {
-	long double factor, busy, total;
+	long double factor, busy = 0, total;
 
 	if (survey->filled & SURVEY_HAS_CHAN_TIME_BUSY)
 		busy = survey->channel_time_busy;
 	else if (survey->filled & SURVEY_HAS_CHAN_TIME_RX)
 		busy = survey->channel_time_rx;
-	else {
-		wpa_printf(MSG_ERROR, "ACS: Survey data missing");
-		return 0;
-	}
 
 	total = survey->channel_time;
 
@@ -420,20 +416,19 @@ static int acs_usable_vht80_chan(struct hostapd_channel_data *chan)
 static int acs_survey_is_sufficient(struct freq_survey *survey)
 {
 	if (!(survey->filled & SURVEY_HAS_NF)) {
+		survey->nf = -95;
 		wpa_printf(MSG_INFO, "ACS: Survey is missing noise floor");
-		return 0;
 	}
 
 	if (!(survey->filled & SURVEY_HAS_CHAN_TIME)) {
+		survey->channel_time = 0;
 		wpa_printf(MSG_INFO, "ACS: Survey is missing channel time");
-		return 0;
 	}
 
 	if (!(survey->filled & SURVEY_HAS_CHAN_TIME_BUSY) &&
 	    !(survey->filled & SURVEY_HAS_CHAN_TIME_RX)) {
 		wpa_printf(MSG_INFO,
 			   "ACS: Survey is missing RX and busy time (at least one is required)");
-		return 0;
 	}
 
 	return 1;
